packages/world/src/systems/MineSystem.sol

```diff
@@ -28,82 +33,171 @@ import { getObjectTypeSchema } from "../utils/ObjectTypeUtils.sol";
+      remaining[i] = max[i] - mined[i];
+      totalRemaining += remaining[i];
+    }
+
+    require(totalRemaining > 0, "No ores available to mine");
+
+    uint256 oreIndex = 0;
+    {
+      // Scale random number to total remaining
+      // TODO: use muldiv from solady or OZ to prevent overflow
+      uint256 scaledRand = (rand * totalRemaining) / type(uint256).max;
+
+      uint256 acc;
+      for (; oreIndex < remaining.length - 1; oreIndex++) {
+        acc += remaining[oreIndex];
+        if (scaledRand < acc) break;
+      }
+    }
+
+    ObjectTypeId ore = ores[oreIndex];
+    MinedOreCount._set(ore, mined[oreIndex] + 1);
+
+    return ore;
+  }
+
+  function processMassReduction(EntityId playerEntityId, EntityId minedEntityId) public returns (uint128) {
+    (uint128 toolMassReduction, ) = useEquipped(playerEntityId);
+    uint128 totalMassReduction = energyToMass(PLAYER_MINE_ENERGY_COST) + toolMassReduction;
+    uint128 massLeft = Mass._getMass(minedEntityId);
+    return massLeft <= totalMassReduction ? 0 : massLeft - totalMassReduction;
+  }
+}
 
 contract MineSystem is System {
   using VoxelCoordLib for *;
 
-  function mineObjectAtCoord(VoxelCoord memory coord) internal returns (EntityId, ObjectTypeId) {
-    require(inWorldBorder(coord), "Cannot mine outside the world border");
+  /// @dev Get relative schema coords, including base coord
+  function _getRelativeCoords(
+    ObjectTypeId objectTypeId,
+    VoxelCoord memory baseCoord
+  ) internal pure returns (VoxelCoord[] memory) {
+    VoxelCoord[] memory schemaCoords = getObjectTypeSchema(objectTypeId);
+    VoxelCoord[] memory coords = new VoxelCoord[](schemaCoords.length + 1);
+
+    coords[0] = baseCoord;
+
+    for (uint256 i = 0; i < schemaCoords.length; i++) {
+      coords[i + 1] = VoxelCoord(
+        baseCoord.x + schemaCoords[i].x,
+        baseCoord.y + schemaCoords[i].y,
+        baseCoord.z + schemaCoords[i].z
+      );
+    }
 
-    (EntityId entityId, ObjectTypeId mineObjectTypeId) = coord.getOrCreateEntity();
-    require(mineObjectTypeId.isBlock(), "Cannot mine non-block object");
-    require(mineObjectTypeId != AnyOreObjectID, "Ore must be computed before it can be mined");
-    require(mineObjectTypeId != AirObjectID, "Cannot mine air");
-    require(mineObjectTypeId != WaterObjectID, "Cannot mine water");
+    return coords;
+  }
 
-    require(entityId.getChipAddress() == address(0), "Cannot mine a chipped block");
-    EnergyData memory machineData = updateMachineEnergyLevel(entityId);
-    require(machineData.energy == 0, "Cannot mine a machine that has energy");
+  function _removeBlock(EntityId entityId, VoxelCoord memory coord) internal {
+    ObjectType._set(entityId, AirObjectID);
 
-    return (entityId, mineObjectTypeId);
+    VoxelCoord memory aboveCoord = VoxelCoord(coord.x, coord.y + 1, coord.z);
+    EntityId aboveEntityId = aboveCoord.getPlayer();
+    if (aboveEntityId.exists()) {
```

Comment:
why was the `&& aboveEntityId.isBaseEntity()` check removed?

I think we need that since only then would gravity apply

Reply:
is it possible for the above player entity not to be the base entity? That would mean the top entity (not base) is directly on top of a block, which from my understanding is not possible

Reply:
true I guess this is not possible cuz players are on top of each other. Was more thinking if we add another type of body like a drone that is 2 wide. But hmm yea dont need the extra read then!

Reply:
sg will add a comment for now

Reply:
actually, if we expect to have other movable objects (being players or others) shouldn't that table be called differently?